lib/libc/minimal/source/stdlib/malloc.c

```diff
@@ -0,0 +1,124 @@
+#include <stdlib.h>
+#include <zephyr.h>
+#include <init.h>
+#include <errno.h>
+#include <misc/mempool.h>
+#include <string.h>
+
+#if (CONFIG_MALLOC_ARENA_SIZE > 0)
+K_MUTEX_DEFINE(malloc_mutex);
+SYS_MEM_POOL_DEFINE(z_malloc_mem_pool, &malloc_mutex,  64,
+		    CONFIG_MALLOC_ARENA_SIZE, 1, 4, .data);
+
+void *malloc(size_t size)
+{
+	void *ret;
+
+	ret = sys_mem_pool_alloc(&z_malloc_mem_pool, size);
+	if (!ret) {
+		errno = ENOMEM;
+	}
+
+	return ret;
+}
+
+static int malloc_prepare(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+#ifdef CONFIG_USERSPACE
+	k_object_access_all_grant(&malloc_mutex);
+#endif
+	sys_mem_pool_init(&z_malloc_mem_pool);
+
+	return 0;
+}
+
+SYS_INIT(malloc_prepare, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+#else /* No malloc arena */
+void *malloc(size_t size)
+{
+	ARG_UNUSED(size);
+
+	errno = ENOMEM;
+
+	return NULL;
+}
+#endif
+
+void free(void *ptr)
+{
+	sys_mem_pool_free(ptr);
+}
+
+static bool size_t_mul_overflow(size_t a, size_t b, size_t *res)
+{
+#if __SIZEOF_SIZE_T__ == 4
+	return __builtin_umul_overflow((unsigned int)a, (unsigned int)b,
+				       (unsigned int *)res);
+#else /* __SIZEOF_SIZE_T__ == 8 */
+	return __builtin_umulll_overflow((unsigned long long)a,
+					 (unsigned long long)b,
+					 (unsigned long long *)res);
+#endif
+}
+
+void *calloc(size_t nmemb, size_t size)
+{
+	if (size_t_mul_overflow(nmemb, size, &size)) {
+		errno = ENOMEM;
+		return NULL;
+	}
+	return malloc(size);
+}
+
+void *realloc(void *ptr, size_t requested_size)
+{
+	struct sys_mem_pool_block *blk;
+	size_t block_size, total_requested_size;
+	void *new_ptr;
+
+	if (requested_size == 0) {
+		return NULL;
+	}
+
+	/* Stored right before the pointer passed to the user */
+	blk = (struct sys_mem_pool_block *)((char *)ptr - sizeof(*blk));
+
```

Comment:
Would be cleaner to expose the needed bits from the mempool API directly, like "sys_mem_pool_get_block(void *ptr)" or whatever.  Probably not a big deal in internal code like this.